fendermint/vm/interpreter/src/chain.rs

```diff
@@ -91,12 +91,19 @@ where
     ) -> anyhow::Result<(Self::State, Self::Output)> {
         match msg {
             ChainMessage::Signed(msg) => {
-                let (state, ret) = self.inner.check(state, *msg, is_recheck).await?;
+                let (state, ret) = self.inner.check(state, msg, is_recheck).await?;
 
                 Ok((state, Ok(ret)))
             }
-            ChainMessage::ForExecution(_) | ChainMessage::ForResolution(_) => {
-                // Users cannot send these messages, only validators can propose them in blocks.
+            ChainMessage::Ipc(IpcMessage::BottomUpResolve(_msg)) => {
+                // TODO: Check the relayer signature and nonce. Don't have to check the quorum certificate, if it's invalid, make the relayer pay.
```

Comment:
That's right, I thought it's just a modified version of the IPC Agent: where it currently has to sign a checkpoint and then send the transaction to the parent, now it has to listen to an event emitted from a smart contract on the child, accumulating the signatures without the help of the agent, and then take that checkpoint, wrap it in a special relay transaction (for CID visibility) and send it to the parent. 

Ostensibly we could send it as an FVM transaction and inspect the `to`, and parse it, instead of having a special IPC transaction :thinking:

Reply:
> take that checkpoint, wrap it in a special relay transaction (for CID visibility) and send it to the parent.

Let's discuss this in today's sync, I am having a hard time understanding the end-to-end flow you are imagining of how the checkpoint signatures are collected, and the actual submission in the parent.